README.md
[![Nuka Carousel — Formidable, We build the modern web](https://raw.githubusercontent.com/FormidableLabs/nuka-carousel/main/Nuka-Hero.png)](https://formidable.com/open-source/)

[![Maintenance Status][maintenance-image]](#maintenance-status)

Small, fast and accessibility-first React carousel library with easily customizable UI and behavior to fit your brand and site

![Nuka Carousel Animated Example](https://i.imgur.com/UwP5gle.gif)

> If you are looking for v4 documentation, you can find it [here](https://www.npmjs.com/package/nuka-carousel/v/4.8.4)

### Install

To add `nuka-carousel` to your project run the following command in your project folder.

```bash
$ yarn add nuka-carousel
```

OR

```bash
$ npm install nuka-carousel
```

> Migration guide from v4 to v5 is under development, please raise any questions in the Issues tab of the repository.

### Example

You can test nuka-carousel default behaviour. [Link](https://nuka-carousel-next.vercel.app/)

```jsx
<Carousel>
  <img src="/image1.png" />
  <img src="/image2.png" />
  <img src="/image3.png" />
  <img src="/image4.png" />
  <img src="/image5.png" />
</Carousel>
```

Infinity nuka-carousel@5 with 3 slides to show. [Link](https://nuka-carousel-next.vercel.app/?slides=6&params=%7B%22wrapAround%22:true,%22slidesToShow%22:3%7D)

```jsx
<Carousel wrapAround={true} slidesToShow={3}>
  <img src="/image1.png" />
  <img src="/image2.png" />
  <img src="/image3.png" />
  <img src="/image4.png" />
  <img src="/image5.png" />
</Carousel>
```

You can play with `&params` url parameter to add or remove any carousel parameters and see how the carousel behaves. We are looking to build a proper documentation page with many examples and code snippets.

### Improvements in v5

- **Support server-side rendering**. Currently in v4 all of the slides are positioned with `position: absolute`, which doesn't render the slides appropriately when javascript is disabled and there are flickering when the front-end is hydrated. This forced users to hardcode properties for the carousel like `initialSlideHeight` and `initialSlideWidth`. In v5 all slides are positioned with dynamic width, so the slides and carousel can be fully responsive and rendered correctly on server as well.
- Rewritten the library with **TypeScript** and **React Hooks** for obvious reasons.
- **Reduce the size of the library and its dependencies**. Currently v4 has 5 dependencies - `prop-types`, `d3-ease`, `wicg-inert`, `exenv` and `react-move`. We are looking to use only `d3-ease` in v5.1, but definitely after bumping the version of it to the latest.
  - `prop-types` is replaced with usage of TypeScript.
  - `exenv` is not maintained anymore and is removed with v5 as a dependency of nuka-carousel.
  - `wicg-inert` doesn't have specified license, so we are going to remove it from our dependencies. [More info here](https://github.com/WICG/inert/issues/168). We are still supporting `inert` and we will have an example and more detailed documentation how you can use it with nuka-carousel v5.
  - `react-move` nothing personal, we just decided to go for our custom approach of animations in order to reduce the size of the library.
- **Fixed issues**. We fixed a lot of the issues that we currently have in v4. You can see the full list with fixed issues in our v5 project. [Link](https://github.com/FormidableLabs/nuka-carousel/projects/1)

### Props

| Name | Type | Description |Default |
| :------------------------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | :----------------------------------------------------------------------------------------------------------------- |
| adaptiveHeight | `boolean`| If it's set to true, the carousel will adapt its height to the visible slides. | `false` |
| adaptiveHeightAnimation | `boolean`| Whether to animate height changes when `adaptiveHeight` is enabled | `true` |
| afterSlide | `(index: number) => void`| Hook to be called after a slide is changed. |`() => {}`|
| animation | `'zoom' \| 'fade'`| Adds a zoom effect on the currently visible slide or change the animation to `fade`. A `transform: scale(0.85)` is set as default when you are using zoom, however, the scale can be customized using `zoomScale` prop. Property is applied on all slides except the current 1. Use `cellAlign` to align the slide with zoom effect where you'd like. ||
| autoplay | `boolean` | Autoplay mode active. | `false`|
| autoplayInterval | `number` | Interval for autoplay iteration in milliseconds. | `3000`  |
| autoplayReverse | `boolean` | Only meaningful when `autoplay` is already true. When `autoplayReverse` is also true, autorotation cycles through slides indexes from high to low. | `false` |
| beforeSlide | `(currentSlideIndex: number, endSlideIndex: number) => void` | Hook to be called before a slide is changed | `() => {}` |
| cellAlign | `'left' \| 'center' \| 'right'` | When displaying more than one slide, sets which position to anchor the current slide to. | `left` |
| cellSpacing | `number` | Space between slides, as an integer, but reflected as `px` | `0` |
| className | `string` | Slider frame class name | `''` |
| defaultControlsConfig | <pre>interface DefaultControlsConfig  { &#13; containerClassName?: string; &#13; nextButtonClassName?: string; &#13; nextButtonOnClick?: (event: React.MouseEvent) => void; &#13; nextButtonStyle?: CSSProperties; &#13; nextButtonText?: React.ReactNode; &#13; pagingDotsClassName?: string; &#13; pagingDotsContainerClassName?: string; &#13; pagingDotsOnClick?: (event: React.MouseEvent) => void; &#13; pagingDotsStyle?: CSSProperties; &#13; prevButtonClassName?: string; &#13; prevButtonOnClick?: (event: React.MouseEvent) => void; &#13; prevButtonStyle?: CSSProperties; &#13; prevButtonText?: React.ReactNode; &#13;}</pre> | This prop lets you apply custom classes and styles to the default `Container`. `Next`, `Previous`, and `Paging Dots` controls. More information on how to customize these controls can be found in the [defaultControlsConfig section](#defaultcontrolsconfig).| `{}` |
| disableAnimation | `boolean` | When set to `true`, will disable animation. | `false` |
| disableEdgeSwiping | `boolean` | When set to `true`, will disable swiping before first slide and after last slide. | `false` |
| dragging | `boolean` | Enable mouse swipe/dragging. | `true` |
| dragThreshold | `number` | The percentage (from 0 to 1) of a slide that the user needs to drag before a slide change is triggered. | `0.5` |
| easing | `(normalizedTime: number) => number` | Animation easing function. See the [Easing section](#easing-and-edgeeasing) for more details | A cubic easeOut function |
| edgeEasing | `(normalizedTime: number) => number` | Animation easing function when swipe exceeds edge. See the [Easing section](#easing-and-edgeeasing) for more details | A cubic easeOut function |
| enableKeyboardControls | `boolean` | When set to `true` will enable keyboard controls when the carousel has focus. If the carousel does not have focus, keyboard controls will be ignored. | `false` |
| frameAriaLabel | `string` | Customize the aria-label of the frame container of the carousel. This is useful when you have more than one carousel on the page. | `''` |
| innerRef | `MutableRefObject<HTMLDivElement>` | React `ref` that should be set on the carousel element | |
| keyCodeConfig | <pre>interface KeyCodeConfig { &#13;  firstSlide?: number[]; &#13;  lastSlide?: number[];&#13;  nextSlide?: number[]; &#13;  pause?: number[]; &#13;  previousSlide?: number[]; &#13;}</pre> | If `enableKeyboardControls` prop is true, you can pass configuration for the keyCode so you can override the default keyboard keys configured. | `{ nextSlide: [39, 68, 38, 87], previousSlide: [37, 65, 40, 83], firstSlide: [81], lastSlide: [69], pause: [32] }` |
| onDragStart | `(e: React.TouchEvent<HTMLDivElement> \| React.MouseEvent<HTMLDivElement>) => void;` | Adds a callback to capture event at the start of swiping/dragging slides | |
| onDrag | `(e: React.TouchEvent<HTMLDivElement> \| React.MouseEvent<HTMLDivElement>) => void;` | Adds a callback to capture swiping/dragging event on slides | |
| onDragEnd | `(e: React.TouchEvent<HTMLDivElement> \| React.MouseEvent<HTMLDivElement>) => void;` | Adds a callback to capture event at the end of swiping/dragging slides | |
| onUserNavigation | `(e: React.TouchEvent \| React.MouseEvent \| React.KeyboardEvent) => void;` | Callback called when user-triggered navigation occurs: dragging/swiping, clicking one of the controls (custom controls not included), or using a keyboard shortcut | |
| pauseOnHover | `boolean` | Pause autoPlay when mouse is over carousel. | `true` |
| renderAnnounceSlideMessage | `(props: Pick<CarouselState, 'currentSlide' \| 'count'>) => string` | Renders message in the ARIA live region that is announcing the current slide on slide change | Render function that returns `"Slide {currentSlide + 1} of {slideCount}"` |
| scrollMode | `'page' \| 'remainder'` | Set `scrollMode="remainder"` if you don't want to see the white space when you scroll to the end of a non-infinite carousel. scrollMode property is ignored when wrapAround is enabled | `'page'` |
| slideIndex | `number` | Manually set the index of the slide to be shown | |
| slidesToScroll | `number` | Slides to scroll at once. The property is overridden to `slidesToShow` when `animation="fade"` | 1 |
| slidesToShow | `number` | Number of slides to show at once. Will be cast to an `integer` when `animation="fade"` | 1 |
| speed | `number` | Animation duration/Transition speed in milliseconds | `500` |
| style  | `CSSProperties` | Add inline style to the carousel frame | `{}` |
| swiping  | `boolean` | Enable touch swipe/dragging | `true` |
| withoutControls | `boolean` | Used to remove all controls at once. Overwrites the `render[Top, Right, Bottom, Left]CenterControls()`. | `false` |
| wrapAround | `boolean` | Sets infinite wrapAround mode. An option similar to repeat or infinite in other libs. | `false` |
| zoomScale | `number` | Adds a number value to set the scale of zoom when `animation === "zoom"`. The number value should be set in a range of (0,1). | `0.85` |


#### render\*Controls

Type: `(props: ControlProps) => ReactElement`

A set of eight render props for rendering controls in different positions around the carousel.

- Valid render props for the eight positions are `renderTopLeftControls`, `renderTopCenterControls`, `renderTopRightControls`, `renderCenterLeftControls`, `renderCenterCenterControls`, `renderCenterRightControls`, `renderBottomLeftControls`, `renderBottomCenterControls`, and `renderBottomRightControls`.

- The default props are set as `renderCenterLeftControls` for `Previous` button, `renderCenterRightControls` for the `Next` button and `renderBottomCenterControls` for the "Paging dots". To change the position or remove "Paging dots", the default positions need to be disabled by setting them to null.

- You can remove all render controls using the `withoutControls` prop on `Carousel`.

- The render functions receive a `ControlProps` argument containing the following props from the Carousel props, using default values if not originally defined: 

  ```
  cellAlign
  cellSpacing
  defaultControlsConfig
  onUserNavigation
  scrollMode
  slidesToScroll
  slidesToShow
  wrapAround
  ```

  Additionally, the following data and callbacks are provided to make creating controls easier:

  | Name                 | Type                            | Description                                             |
  | :------------------- | ------------------------------- | :------------------------------------------------------ |
  | currentSlide         | `number`                        | Current slide index                                     |
  | pagingDotsIndices    | `number[]`                      | The indices for the paging dots                         |
  | goToSlide            | `(targetIndex: number) => void` | Go to a specific slide                                  |
  | nextDisabled         | `boolean`                       | Whether the "next" button should be disabled or not     |
  | nextSlide            | `() => void`                    | Go to the next slide                                    |
  | previousDisabled     | `boolean`                       | Whether the "previous" button should be disabled or not |
  | previousSlide        | `() => void`                    | Go to the previous slide                                |
  | slideCount           | `number`                        | Total number of slides                                  |

Example:

```jsx
<Carousel
  renderTopCenterControls={({ currentSlide }) => (
    <div>Slide: {currentSlide}</div>
  )}
  renderCenterLeftControls={({ previousDisabled, previousSlide }) => (
    <button onClick={previousSlide} disabled={previousDisabled}>
      Previous
    </button>
  )}
  renderCenterRightControls={({ nextDisabled, nextSlide }) => (
    <button onClick={nextSlide} disabled={nextDisabled}>
      Next
    </button>
  )}
>
  {/* Carousel Content */}
</Carousel>
```

- NOTE: The className `slide-visible` is added to the currently visible slide or slides (when `slidesToShow` > 1). The className `slide-current` is added to the currently "active" slide.

#### easing and edgeEasing

`(normalizedTime: number) => number`

A function accepting a normalized time between 0 and 1, inclusive, and returning an eased time, which equals 0 at normalizedTime==0 and equals 1 at normalizedTime==1. You can plug in your own custom easing function (e.g., `(t) => t` for a linear transition), or import functions from a different library, like [`d3-ease`](https://github.com/d3/d3-ease).
```jsx
import { easeCircleOut, easeElasticOut } from 'd3-ease';

// ...

<Carousel easing={easeCircleOut} edgeEasing={easeElasticOut}>
  {/* Carousel Content */}
</Carousel>
```

Please note that using a function for `easing` with "In" in it (ease**In**Out, easeElastic**In**, etc.) will make swiping transitions feel a bit clunky, as the velocity at the end of the swipe will suddenly drop to follow the slow startup speed of the "In" easing function. In general, when using custom easing functions, try out both swiping and clicking on the navigation buttons to see how the transitions feel.

#### renderAnnounceSlideMessage

`(props: Pick<CarouselState, 'currentSlide' \| 'count'>) => string`

`renderAnnounceSlideMessage` render prop is a special case of the `render*Controls` props. It's responsibility is to render ARIA live announcement message to improve accessibility. The prop will announce the message you pass in every time the slide changes with `VoiceOver` enabled on your machine. The function returns only `slideCount` and `currentSlide` values.

```jsx
<Carousel
  renderAnnounceSlideMessage={({ currentSlide, slideCount }) =>
    `Slide ${currentSlide + 1} of ${slideCount}`
  }
>
  {/* Carousel Content */}
</Carousel>
```

#### defaultControlsConfig

```
interface DefaultControlsConfig {
  containerClassName?: string;
  nextButtonClassName?: string;
  nextButtonOnClick?: (event: React.MouseEvent) => void;
  nextButtonStyle?: CSSProperties;
  nextButtonText?: React.ReactNode;
  pagingDotsClassName?: string;
  pagingDotsContainerClassName?: string;
  pagingDotsOnClick?: (event: React.MouseEvent) => void;
  pagingDotsStyle?: CSSProperties;
  prevButtonClassName?: string;
  prevButtonOnClick?: (event: React.MouseEvent) => void;
  prevButtonStyle?: CSSProperties;
  prevButtonText?: React.ReactNode;
}
```

The default controls used by Nuka are the `Previous` button, `Next` button, and `PagingDots` control. The visual look and text of these controls can be modified with props as described below:

- The props ending with `ClassName` let you apply a custom css class to its respective control.
- The props ending with `Style` let you apply inline styles to its respective control.
- The props ending with `OnClick` let you listen for user interaction with the controls.
- The text label for the `Previous` button and `Next` button can be customized using `prevButtonText` and `nextButtonText`, respectively.

For example, you can change the text of the `Previous` and `Next` buttons, and change the paging dots to red by passing in the following configuration:

```
defaultControlsConfig={{
  nextButtonText: 'Custom Next',
  prevButtonText: 'Custom Prev',
  pagingDotsStyle: {
    fill: 'red'
  }
}}
```

<!-- ### External Control of Carousel State

You can control the state of the carousel from your parent component as shown below:

```jsx
import React from 'react';
import Carousel from 'nuka-carousel';

export default class extends React.Component {
  state = {
    slideIndex: 0
  };

  render() {
    return (
      <Carousel
        slideIndex={this.state.slideIndex}
        afterSlide={(slideIndex) => this.setState({ slideIndex })}
      >
        <img src="https://via.placeholder.com/400/ffffff/c0392b/&text=slide1" />
        <img src="https://via.placeholder.com/400/ffffff/c0392b/&text=slide2" />
        <img src="https://via.placeholder.com/400/ffffff/c0392b/&text=slide3" />
        <img src="https://via.placeholder.com/400/ffffff/c0392b/&text=slide4" />
        <img src="https://via.placeholder.com/400/ffffff/c0392b/&text=slide5" />
        <img src="https://via.placeholder.com/400/ffffff/c0392b/&text=slide6" />
      </Carousel>
    );
  }
}
``` -->

### Contributing

See the [Contribution Docs](CONTRIBUTING.md).

### Maintenance Status

**Active:** Formidable is actively working on this project, and we expect to continue for work for the foreseeable future. Bug reports, feature requests and pull requests are welcome.

[maintenance-image]: https://img.shields.io/badge/maintenance-active-green.svg?color=brightgreen&style=flat


CHANGELOG.md
# Nuka Changelog

## 5.5.1

### Patch Changes

* Fix to only include inert property when slide is not in view ([#1000](https://github.com/FormidableLabs/nuka-carousel/pull/1000)] @allypalanzi


## 5.5.0

### Minor Changes

* Implement fixed width slides with intersection observer ([#985](https://github.com/FormidableLabs/nuka-carousel/pull/985))
* Fix cellAlignment issue
* Enforce `scrollMode` `remainder` if `slidesToScroll` `auto` is used

## 5.4.1

### Patch Changes

- slider-frame is no longer a focus target if enableKeyboardControls is false ([#984](https://github.com/FormidableLabs/nuka-carousel/pull/984))

* fix the Alignment type so library users don't need to import an enum for a simple value ([#974](https://github.com/FormidableLabs/nuka-carousel/pull/974))

## 5.4.0

### Minor Changes

- add onUserNavigation prop to listen for user-triggered navigation ([#970](https://github.com/FormidableLabs/nuka-carousel/pull/970))

* allow for any element or component in the nextButtonText and prevButtonText props, instead of just strings ([#970](https://github.com/FormidableLabs/nuka-carousel/pull/970))

- pass nextDisabled, previousDisabled, and pagingDotsIndices to render\*Controls callbacks to aid in the creation of custom controls ([#966](https://github.com/FormidableLabs/nuka-carousel/pull/966))

* adds the ability to use custom easing functions for the animations via `easing` and `edgeEasing` ([#959](https://github.com/FormidableLabs/nuka-carousel/pull/959))

- add props to enable users to hook into interactions with the default carousel controls ([#970](https://github.com/FormidableLabs/nuka-carousel/pull/970))

### Patch Changes

- animation=fade now behaves as it did pre-5.3.0, scrolling all visible slides at once ([#959](https://github.com/FormidableLabs/nuka-carousel/pull/959))

* make the control dots change in size when dimensions changed via CSS ([#964](https://github.com/FormidableLabs/nuka-carousel/pull/964))

- Applies draggable=false only to children of sliderList, not entire document ([#972](https://github.com/FormidableLabs/nuka-carousel/pull/972))

## 5.3.0

### Minor Changes

- makes it possible to swipe over multiple slides at once rather than being limited to one ([#956](https://github.com/FormidableLabs/nuka-carousel/pull/956))

### Patch Changes

- Fix autoplay timing so the interval is not extended occasionally when wrapAround=true ([#954](https://github.com/FormidableLabs/nuka-carousel/pull/954))

* make the prev/next buttons jump to appropriate indices with cellAlign=center|right ([#952](https://github.com/FormidableLabs/nuka-carousel/pull/952))

- keyboard-triggered "firstSlide" or "lastSlide" actions are now animated, and now take cellAlign into account ([#958](https://github.com/FormidableLabs/nuka-carousel/pull/958))

* make autoplay continue to the last slide when cellAlign != left and slidesToShow > 1 ([#952](https://github.com/FormidableLabs/nuka-carousel/pull/952))

- children of Carousel that are falsy will not be rendered as slides ([#953](https://github.com/FormidableLabs/nuka-carousel/pull/953))

* fix missing slide-visible classes when slideIndex has fractional digits ([#947](https://github.com/FormidableLabs/nuka-carousel/pull/947))

- correct button disabling when cellAlign=center|right ([#952](https://github.com/FormidableLabs/nuka-carousel/pull/952))

* show a dot corresponding to the last slide when cellAlign=center|right ([#952](https://github.com/FormidableLabs/nuka-carousel/pull/952))

- fixes number of dots in default controls and eliminates janky animation when changing slides rapidly ([#945](https://github.com/FormidableLabs/nuka-carousel/pull/945))

* setting "dragging" to false will no longer disable carousel swiping on mobile ("swiping" does that) ([#956](https://github.com/FormidableLabs/nuka-carousel/pull/956))

- keyboard interactions when focused on the carousel are now kept from bubbling up and triggering other listeners ([#958](https://github.com/FormidableLabs/nuka-carousel/pull/958))

* fix initial index when autoplayReverse=true and cellAlign is not left ([#952](https://github.com/FormidableLabs/nuka-carousel/pull/952))

## 4.8.3 (2021-11-19)

- [#810](https://github.com/FormidableLabs/nuka-carousel/pull/810) fix: Last image is not rendered with 3 slides

## 4.8.2 (2021-10-29)

- [#808](https://github.com/FormidableLabs/nuka-carousel/pull/808) update webpack deps
- [#807](https://github.com/FormidableLabs/nuka-carousel/pull/807) Fix: improve pinch zoom behavior on mobile

## 4.8.1 (2021-10-22)

- [#804](https://github.com/FormidableLabs/nuka-carousel/pull/804) Fix: flip inequality check in initializeCarouselHeight
- [#805](https://github.com/FormidableLabs/nuka-carousel/pull/805) Fix: prevent initializeCarouselHeight loop to fix browser crash

## 4.8.0 (2021-09-21)

- [#802](https://github.com/FormidableLabs/nuka-carousel/pull/802) Minor A11y improvements
- [#801](https://github.com/FormidableLabs/nuka-carousel/pull/801) Fix lint errors
- [#800](https://github.com/FormidableLabs/nuka-carousel/pull/800) Dep bumps
- [#799](https://github.com/FormidableLabs/nuka-carousel/pull/799) Dep bumps
- [#796](https://github.com/FormidableLabs/nuka-carousel/pull/796) Perf: Use more modern screen-reader only styles
- [#794](https://github.com/FormidableLabs/nuka-carousel/pull/794) Dep bumps

## 4.7.9 (2021-07-13)

- [#791](https://github.com/FormidableLabs/nuka-carousel/pull/791) Fix: bind blockEvent method to `this`
- [#785](https://github.com/FormidableLabs/nuka-carousel/pull/785) Fix: update react-move
- [#781](https://github.com/FormidableLabs/nuka-carousel/pull/781) Fix: Improve SVG dot A11y
- [#779](https://github.com/FormidableLabs/nuka-carousel/pull/779) Dep bumps
- [#777](https://github.com/FormidableLabs/nuka-carousel/pull/777)
- [#776](https://github.com/FormidableLabs/nuka-carousel/pull/776)

## 4.7.8 (2021-05-10)

- [#766](https://github.com/FormidableLabs/nuka-carousel/pull/766) Dep bumps

index.d.ts
import * as React from 'react';

export type CarouselScrollModeProp = 'page' | 'remainder';

export type CarouselCellAlignProp = 'left' | 'center' | 'right';

export type CarouselSlidesToScrollProp = number | 'auto';

export type CarouselSlideWidthProp = string | number;

export type CarouselSlideActions =
  | 'nextSlide'
  | 'previousSlide'
  | 'firstSlide'
  | 'lastSlide'
  | 'pause';

export type CarouselControlContainerProp =
  | 'TopLeft'
  | 'TopCenter'
  | 'TopRight'
  | 'CenterLeft'
  | 'CenterCenter'
  | 'CenterRight'
  | 'BottomLeft'
  | 'BottomCenter'
  | 'BottomRight';

export interface CarouselSlideRenderControlProps {
  /**
   * When displaying more than one slide, sets which position to anchor the current slide to.
   */
  cellAlign: CarouselCellAlignProp;

  /**
   * Space between slides, as an integer, but reflected as px
   */
  cellSpacing: number;

  /**
   * Current slide index
   */
  currentSlide: number;

  /**
   * Current frame width (px)
   */
  frameWidth: number;

  /**
   * Go to X slide method
   * @param index Slide's index to go
   */
  goToSlide: (index: number) => void;

  /**
   * Go to the next slide method
   */
  nextSlide: () => void;

  /**
   * Go to the previous slide method
   */
  previousSlide: () => void;

  /**
   * Total amount of slides
   */
  slideCount: number;

  /**
   * Slides to scroll at once
   */
  slidesToScroll: CarouselSlidesToScrollProp;

  /**
   * Slides to show at once
   */
  slidesToShow: number;

  /**
   * Current slide width (px)
   */
  slideWidth: number;

  /**
   * Infinite mode enabled
   */
  wrapAround: boolean;
}

export type CarouselRenderControl = (
  props: CarouselSlideRenderControlProps
) => React.ReactNode;

export interface CarouselProps {
  /**
   * If it's set to true, the carousel will adapt its height to the visible slides.
   */
  adaptiveHeight?: boolean;

  /**
   * Hook to be called after a slide is changed
   * @param index Index of the current slide
   */
  afterSlide?: (index: number) => void;

  /**
   * Adds a zoom or fade effect on the currently visible slide.
   */
  animation?: 'zoom' | 'fade';

  /**
   * Autoplay mode active
   * @default false
   */
  autoplay?: boolean;

  /**
   * Interval for autoplay iteration (ms)
   * @default 3000
   */
  autoplayInterval?: number;

  /**
   * Autoplay cycles through slide indexes in reverse
   * @default false
   */
  autoplayReverse?: boolean;

  /**
   * Hook to be called before a slide is changed
   * @param currentSlide Index of the current slide
   * @param endSlide Index of the last slide
   */
  beforeSlide?: (currentSlide: number, endSlide: number) => void;

  /**
   * When displaying more than one slide,
   * sets which position to anchor the current slide to
   */
  cellAlign?: CarouselCellAlignProp;

  /**
   * Space between slides, as an integer, but reflected as px
   */
  cellSpacing?: number;

  /**
   * Explicit children prop to resolve issue with @types/react v18
   */
  children?: React.ReactNode;

  /**
   * Additional className
   */
  className?: string;

  /**
   * This prop lets you apply custom classes and styles to the default Next, Previous, and Paging Dots controls
   */
  defaultControlsConfig?: {
    containerClassName?: string;
    nextButtonClassName?: string;
    nextButtonStyle?: React.CSSProperties;
    nextButtonText?: string;
    prevButtonClassName?: string;
    prevButtonStyle?: React.CSSProperties;
    prevButtonText?: string;
    pagingDotsContainerClassName?: string;
    pagingDotsClassName?: string;
    pagingDotsStyle?: React.CSSProperties;
  };

  /**
   * Disable slides animation
   * @default false
   */
  disableAnimation?: boolean;

  /**
   * Disable swipe before first slide and after last slide
   * @default false
   */
  disableEdgeSwiping?: boolean;

  /**
   * Enable mouse swipe/dragging
   */
  dragging?: boolean;

  /**
   * The percentage (from 0 to 1) of a slide that the user needs to drag before a slide change is triggered.
   */
  dragThreshold?: number;

  /**
   * Animation easing function
   * @see https://github.com/d3/d3-ease
   */
  easing?: string;

  /**
   * Animation easing function when swipe exceeds edge
   * @see https://github.com/d3/d3-ease
   */
  edgeEasing?: string;

  /**
   * When set to true, disable keyboard controls
   * @default false
   */
  enableKeyboardControls?: boolean;

  /**
   * Customize the aria-label of the frame container of the carousel. This is useful when you have more than one carousel on the page.
   */
  frameAriaLabel?: string;

  /**
   * Ref to pass to carousel element
   */
  innerRef?: React.RefObject<HTMLInputElement>;

  /**
   * When enableKeyboardControls is enabled, Configure keyCodes for corresponding slide actions as array of keyCodes
   */
  keyCodeConfig?: {
    [slideAction in CarouselSlideActions]?: number[];
  };

  /**
   * optional callback function
   */
  onDragStart?: (
    e?: React.TouchEvent<HTMLDivElement> | React.MouseEvent<HTMLDivElement>
  ) => void;

  /**
   * optional callback function
   */
  onDrag?: (
    e?: React.TouchEvent<HTMLDivElement> | React.MouseEvent<HTMLDivElement>
  ) => void;

  /**
   * optional callback function
   */
  onDragEnd?: (
    e?: React.TouchEvent<HTMLDivElement> | React.MouseEvent<HTMLDivElement>
  ) => void;

  /**
   * Pause autoPlay when mouse is over carousel
   * @default true
   */
  pauseOnHover?: boolean;

  /**
   * Function for rendering top left control
   */
  renderTopLeftControls?: CarouselRenderControl | null;

  /**
   * Function for rendering top center control
   */
  renderTopCenterControls?: CarouselRenderControl | null;

  /**
   * Function for rendering top right control
   */
  renderTopRightControls?: CarouselRenderControl | null;

  /**
   * Function for rendering center left control
   */
  renderCenterLeftControls?: CarouselRenderControl | null;

  /**
   * Function for rendering center center control
   */
  renderCenterCenterControls?: CarouselRenderControl | null;

  /**
   * Function for rendering center right control
   */
  renderCenterRightControls?: CarouselRenderControl | null;

  /**
   * Function for rendering bottom left control
   */
  renderBottomLeftControls?: CarouselRenderControl | null;

  /**
   * Function for rendering bottom center control
   */
  renderBottomCenterControls?: CarouselRenderControl | null;

  /**
   * Function for rendering bottom right control
   */
  renderBottomRightControls?: CarouselRenderControl | null;

  /**
   * Function for rendering aria-live announcement messages
   */
  renderAnnounceSlideMessage?: ({
    currentSlide,
    slideCount
  }: CarouselSlideRenderControlProps) => string;

  /**
   * Supports 'page' and 'remainder' scroll modes.
   * Defaults to 'remainder'.
   */
  scrollMode?: CarouselScrollModeProp;

  /**
   * Manually set the index of the slide to be shown
   */
  slideIndex?: number;

  /**
   * Slides to scroll at once.
   */
  slidesToScroll?: CarouselSlidesToScrollProp;

  /**
   * Slides to show at once
   */
  slidesToShow?: number;

  /**
   * Animation duration
   */
  speed?: number;

  /**
   * style object
   */
  style?: React.CSSProperties;

  /**
   * Enable touch swipe/dragging
   */
  swiping?: boolean;

  /**
   * Enable the slides to transition vertically
   */
  vertical?: boolean;

  /**
   * Sets infinite wrapAround mode
   * @default false
   */
  wrapAround?: boolean;

  /**
   * Used to remove all controls at once. Overwrites the render[Top, Right, Bottom, Left]CenterControls()
   * @default false
   */
  withoutControls?: boolean;

  /**
   * Adds a number value to set the scale of zoom when animation === "zoom".
   * The number value should be set in a range of (0,1).
   * @default 0.85
   */
  zoomScale?: number;
}

export interface CarouselState {
  /**
   * Current slide index
   */
  currentSlide: number;

  /**
   * Is dragging enabled
   */
  dragging: boolean;

  /**
   * Easing function name
   */
  easing: string;

  isWrappingAround: boolean;

  pauseOnHover: boolean;

  resetWrapAroundPosition: boolean;

  /**
   * Total amount of slides
   */
  slideCount: number;

  /**
   * Slides to scroll at once
   */
  slidesToScroll: CarouselSlidesToScrollProp;

  /**
   * Current slide width
   */
  slideWidth: CarouselSlideWidthProp;

  /**
   * Is infinite mode enabled
   */
  wrapToIndex: boolean;
}

export default class Carousel extends React.Component<
  CarouselProps,
  CarouselState
> {}

export type PreviousButtonProps = CarouselSlideRenderControlProps;
export const PreviousButton: React.FC<PreviousButtonProps>;

export type NextButtonProps = CarouselSlideRenderControlProps;
export const NextButton: React.FC<NextButtonProps>;

export type PagingDotsProps = CarouselSlideRenderControlProps;
export const PagingDots: React.FC<PagingDotsProps>;
